File: csharp/v7/Go/CSharpParserBase.go
package parser

import (
	"os"
	"strings"

	"github.com/antlr4-go/antlr/v4"
)

var allSemanticFunctions = []string{"IsLocalVariableDeclaration"}

type CSharpParserBase struct {
	*antlr.BaseParser
	noSemantics map[string]bool
	initialized bool
}

func (p *CSharpParserBase) initSemantics() {
	if p.initialized {
		return
	}
	p.initialized = true
	p.noSemantics = make(map[string]bool)
	for _, arg := range os.Args {
		lower := strings.ToLower(arg)
		if strings.HasPrefix(lower, "--no-semantics") {
			eq := strings.Index(arg, "=")
			if eq == -1 {
				for _, f := range allSemanticFunctions {
					p.noSemantics[f] = true
				}
			} else {
				for _, f := range strings.Split(arg[eq+1:], ",") {
					p.noSemantics[strings.TrimSpace(f)] = true
				}
			}
		}
	}
}

func (p *CSharpParserBase) IsLocalVariableDeclaration() bool {
	p.initSemantics()
	if p.noSemantics["IsLocalVariableDeclaration"] {
		return true
	}
	ctx := p.GetParserRuleContext()
	local_var_decl, ok := ctx.(*Local_variable_declarationContext)
	if !ok || local_var_decl == nil {
		return true
	}
	local_variable_type := local_var_decl.Local_variable_type()
	if local_variable_type == nil {
		return true
	}
	if local_variable_type.GetText() == "var" {
		return false
	}
	return true
}

File: csharp/v7/Python3/CSharpParserBase.py
import sys
from antlr4 import Parser

ALL_SEMANTIC_FUNCTIONS = ["IsLocalVariableDeclaration"]


def parseNoSemantics(args):
    result = set()
    for a in args:
        if a.lower().startswith("--no-semantics"):
            eq = a.find('=')
            if eq == -1:
                for f in ALL_SEMANTIC_FUNCTIONS:
                    result.add(f)
            else:
                for f in a[eq + 1:].split(','):
                    result.add(f.strip())
    return result


class CSharpParserBase(Parser):

    def __init__(self, input, output=sys.stdout):
        super().__init__(input, output)
        self._noSemantics = parseNoSemantics(sys.argv)

    def IsLocalVariableDeclaration(self):
        if "IsLocalVariableDeclaration" in self._noSemantics:
            return True
        if "." in __name__:
            from .CSharpParser import CSharpParser
        else:
            from CSharpParser import CSharpParser
        local_var_decl = self._ctx
        if not isinstance(local_var_decl, CSharpParser.Local_variable_declarationContext):
            return True
        local_variable_type = local_var_decl.local_variable_type()
        if local_variable_type is None:
            return True
        if local_variable_type.getText() == "var":
            return False
        return True
